# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/vm/StringType.h
# Commit: fabb8df993fb
# Full Hash: fabb8df993fb044dbf52c3c74f6d679cd42c7f73
# Author: Jon Coppeard <jcoppeard@mozilla.com>
# Date: 2021-03-17 21:25:27
# Regressor Bug: 1698543
# File Overlap Count: 4
# Description:
#   Bug 1698543 - Rework handling of interaction between incremental barriers and rope flattening r=sfink
#   
#   Currently we have a special path for performing barriers during rope flattening to take account of the fact that this overwrites the cell header word of ropes temporarily, making these cells untraceable.
#   
#   The problem is that a rope can already be in the barrier buffer before flattening starts and so we need to check for this when processing the buffer.
# ==============================================================================

diff -r 181aeda5c83a -r fabb8df993fb js/src/vm/StringType.h
--- a/js/src/vm/StringType.h	Wed Mar 17 08:50:18 2021 +0000
+++ b/js/src/vm/StringType.h	Wed Mar 17 09:01:16 2021 +0000
@@ -405,11 +405,17 @@
 #endif
   }
 
- protected:
-  void setFlattenData(uintptr_t data) { setTemporaryGCUnsafeData(data); }
+  void setFlattenData(JSString* parent, uintptr_t flags) {
+    MOZ_ASSERT((uintptr_t(parent) & RESERVED_MASK) == 0);
+    MOZ_ASSERT((flags & ~RESERVED_MASK) == 0);
+    setTemporaryGCUnsafeData(uintptr_t(parent) | flags);
+  }
 
-  uintptr_t unsetFlattenData(uint32_t len, uint32_t flags) {
-    return unsetTemporaryGCUnsafeData(len, flags);
+  JSString* unsetFlattenData(uint32_t len, uint32_t newFlags,
+                             uintptr_t* oldFlagsOut) {
+    uintptr_t data = unsetTemporaryGCUnsafeData(len, newFlags);
+    *oldFlagsOut = data & RESERVED_MASK;
+    return reinterpret_cast<JSString*>(data & ~RESERVED_MASK);
   }
 
   // Get correct non-inline chars enum arm for given type
@@ -592,6 +598,8 @@
   mozilla::Maybe<mozilla::Tuple<size_t, size_t>> encodeUTF8Partial(
       const JS::AutoRequireNoGC& nogc, mozilla::Span<char> buffer) const;
 
+  bool isBeingFlattened() const { return hasTempHeaderData(); }
+
  private:
   // To help avoid writing Spectre-unsafe code, we only allow MacroAssembler
   // to call the method below.
